### ISSUE TITLE
Finite Scalar Quantization (not a real issue)

### ISSUE THREAD

**Issue by sl33pyC01E** [madebyollin/taesd#36]
https://arxiv.org/abs/2309.15505

I'm working on something like tiny latent predictor networks, mostly using RSSMs of various flavors along with pretrained encoder/decoder models of various flavors. I haven't found very many instance examples of projects using this approach, most rssm predictors I know of either train a cnn end to end or use a chonky diffusion block for reconstruction quality. I recently found the finite scalar quantization method and achieved honestly shocking results when applying it to your TAESD model (convergence in ~1000 steps at batch16 with negligible quality loss using the codebook plan from open magvit). 

I recently started looking at your other works, and I think the nearest neighbor collapse that occurs in regression for vq-gan type decoders is actually helpful in regard to image sequence stability. I just felt like sharing and saying I'm a huge fan

**Comment by madebyollin** [comment]
Interesting! Are you the same sl33py who was working on Elden Ring emulation? I haven't messed with discrete bottlenecks much, but my understanding is that the KL noising + implicit (model architecture) regularization in continuous VAE bottlenecks makes them pretty robust to low-bitrate quantization, so e.g. https://huggingface.co/datasets/cloneofsimo/imagenet.int8 (8 bits per element) is totally fine.

Regarding temporal stability w/ different bottlenecks:
1. I think it's common to add some noise to the past latents in order to reduce the sensitivity of continuous WMs to past errors (GameNGen https://arxiv.org/abs/2408.14837 does this; my [example LDM WM code](https://github.com/madebyollin/game-emulation-via-neural-network-dataset/blob/main/Example_Diffusion_World_Model.ipynb) does as well). However, there's still potential for drift here.
2. Microsoft's discrete-AR Quake demo (https://www.microsoft.com/en-us/research/articles/whamm-real-time-world-modelling-of-interactive-environments/ / https://copilot.microsoft.com/gaming) is really drift-free, so that might be evidence in favor of explicit quantization.

**Comment by sl33pyC01E** [comment]
I am that Sl33py yes

I've been using added noise as well as substituting output frames in as ground truth frames on a very sparse ramp up schedule, both have improved stability

Currently I am working on adding the Diamond world model architecture to my project as well as looking for datasets and benchmarks to ablate against (I finally broke and couldn't take fighting/looking at margit anymore...)

I'm also looking into quantizing the flux tae down to a discrete 4 channel but haven't started yet and fear it might harm latency too much. Currently I have had the best results with vae -> deterministic RSSM (cut the stochastic block out of DreamerV3) -> Diamond (without the dynamics block) -> vae. The RSSM acts as the dynamics block and the Diamond 'refiner' acts as the stochastic block as well as the visual last mile. Everything fits together pretty nicely during autoregressive rollout (extremely strong results with grounded N steps ahead prediction) but I need to build towards using standardized data and benchmarks before I can say it's any better with confidence. The pretrained discretized vae route does seem to cut data sampling requirements by a lot compared to what I have compared against though

I'll look at those resources you linked! I appreciate the response
